crates/markdown_preview/src/markdown_preview_view.rs

```diff
@@ -0,0 +1,169 @@
+                let contents = editor.buffer().read(cx).snapshot(cx).text();
+                this.contents = contents;
+                cx.notify();
+            }
+        })
+        .detach();
+
+        let editor = active_editor.read(cx);
+        let contents = editor.buffer().read(cx).snapshot(cx).text();
+
+        Self {
+            focus_handle,
+            languages,
+            contents,
+        }
+    }
+
+    // Re-active the most recently active preview, or create a new one if there isn't one.
+    fn deploy_preview(
+        workspace: &mut Workspace,
+        languages: Arc<LanguageRegistry>,
+        active_editor: View<Editor>,
+        cx: &mut ViewContext<Workspace>,
+    ) {
+        let existing = workspace
+            .active_pane()
+            .read(cx)
+            .items()
+            .find_map(|item| item.downcast::<MarkdownPreviewView>());
+
+        Self::existing_or_new_preview(workspace, languages, active_editor, existing, cx)
+    }
+
+    fn existing_or_new_preview(
+        workspace: &mut Workspace,
+        languages: Arc<LanguageRegistry>,
+        active_editor: View<Editor>,
+        existing: Option<View<MarkdownPreviewView>>,
+        cx: &mut ViewContext<Workspace>,
+    ) {
+        if let Some(existing) = existing {
+            workspace.activate_item(&existing, cx);
+            existing
+        } else {
+            let view: View<MarkdownPreviewView> =
+                cx.new_view(|cx| MarkdownPreviewView::new(active_editor, languages, cx));
+
+            workspace.split_item(workspace::SplitDirection::Right, Box::new(view.clone()), cx);
+            view
+        };
+    }
+}
+
+impl FocusableView for MarkdownPreviewView {
+    fn focus_handle(&self, _: &AppContext) -> gpui::FocusHandle {
+        self.focus_handle.clone()
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum PreviewEvent {}
+
+impl EventEmitter<PreviewEvent> for MarkdownPreviewView {}
+
+impl Item for MarkdownPreviewView {
+    type Event = PreviewEvent;
+
+    fn tab_content(
+        &self,
+        _detail: Option<usize>,
+        selected: bool,
+        _cx: &WindowContext,
+    ) -> AnyElement {
+        h_flex()
+            .gap_2()
+            .child(Icon::new(IconName::FileDoc).color(if selected {
+                Color::Default
+            } else {
+                Color::Muted
+            }))
+            .child(Label::new("Markdown preview").color(if selected {
+                Color::Default
+            } else {
+                Color::Muted
+            }))
+            .into_any()
+    }
+
+    fn telemetry_event_text(&self) -> Option<&'static str> {
+        Some("markdown preview")
+    }
+
+    fn to_item_events(_event: &Self::Event, _f: impl FnMut(workspace::item::ItemEvent)) {}
+}
+
+impl Render for MarkdownPreviewView {
+    fn render(&mut self, cx: &mut ViewContext<Self>) -> impl IntoElement {
+        // TODO: This is wrong but I'm unsure how
```

Comment:
This is one of the worst place in gpui for me too, here's something in the same `syntax_tree_view` that might help: https://github.com/zed-industries/zed/blob/843919dd1448684b308977e7b3855f853e9afe8f/crates/language_tools/src/syntax_tree_view.rs#L361-L369

There, we wrap an element in something that has to have its bounds calculated, so we have a way to keep it generic and dynamic.

Reply:
Made these changes, thank you. 

> we wrap an element in something that has to have its bounds calculated

Am I understanding this right?

1. My `render` function returns a `canvas` with `height=100%` and `width=100%`.
2. Taffy lays out the `canvas` appropriately.
3. When GPUI goes to paint the canvas element, another layout happens for my markdown contents (`paint_callback`).
4. Another paint is triggered (a frame later?)

Reply:
I believe you're right except that both canvas' and element's layouting and paints happen at once, no "a frame later" happens for any paint/delay otherwise we would see it blinking on resize.